### PR TITLE
fix: Add key vault to storage account - `avm/res/storage/storage-account`

### LINKED ISSUE #1780: [AVM Question/Feedback]: Accessing Storage Account Access Key
### Check for previous/existing GitHub issues

- [X] I have checked for previous/existing GitHub issues

### Description

We've utilized the AVM approach for deploying several storage accounts, as detailed in the repository: https://github.com/Azure/bicep-registry-modules/tree/main/avm/res/storage/storage-account

What's the recommended method for accessing the access key of these storage accounts? We've observed that the outputs don't include any parameters for this purpose. 

This concern arises particularly when deploying multiple resources, such as storage accounts and Azure Functions, where the Azure Functions' configuration settings needs the connection string containing the access key of the storage account. Currently, we're blocked from achieving this. 

Any guidance on this matter would be greatly appreciated.

### PR DESCRIPTION
**Pull request by Agazoth**
## Pipeline Reference

| Pipeline |
| -------- |
|   [avm.res.storage.storage-account](https://github.com/Agazoth/bicep-registry-modules/actions/workflows/avm.res.storage.storage-account.yml) |

## Type of Change

- [ ] Update to CI Environment or utlities (Non-module effecting changes)
- [ ] Azure Verified Module updates:
  - [ ] Bugfix containing backwards compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [x] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

File: avm/res/document-db/database-account/gremlin-database/graph/main.json
Comment (eriqua) [inline avm/res/document-db/database-account/gremlin-database/graph/main.json:1]: Please keep one PR mapped to a single module, to allow proper review and approval by the right module owner

File: avm/res/storage/storage-account/blob-service/main.bicep [hunk 1]
```diff
@@ -12,7 +12,7 @@ param automaticSnapshotPolicyEnabled bool = false
 @description('Optional. The blob service properties for change feed events. Indicates whether change feed event logging is enabled for the Blob service.')
 param changeFeedEnabled bool = false
 
-@minValue(0)
+@minValue(1)
 @maxValue(146000)
 @description('Optional. Indicates whether change feed event logging is enabled for the Blob service. Indicates the duration of changeFeed retention in days. A "0" value indicates an infinite retention of the change feed.')
```
Comment (AlexanderSehr) [inline avm/res/storage/storage-account/blob-service/main.bicep:17]: Please be aware, the parameter's description & new `min` value don't align.

Comment (Agazoth) [reply avm/res/storage/storage-account/blob-service/main.bicep:17]: I agree, description should be updated to align with the documentation: https://learn.microsoft.com/en-us/azure/templates/microsoft.storage/storageaccounts/blobservices?pivots=deployment-language-bicep#changefeed

File: avm/res/document-db/database-account/gremlin-database/graph/main.json
Comment (Agazoth) [reply avm/res/document-db/database-account/gremlin-database/graph/main.json:1]: I reverted the faulty merged files. Should be ok now.

File: avm/res/storage/storage-account/blob-service/main.bicep (hunk 1 above)
Comment (Agazoth) [reply avm/res/storage/storage-account/blob-service/main.bicep]: Description updated

File: avm/res/document-db/database-account/gremlin-database/graph/main.json
Comment (AlexanderSehr) [reply avm/res/document-db/database-account/gremlin-database/graph/main.json:1]: Please note, I'll close and re-open the PR to have it re-evaulate the rules.

Comment (AlexanderSehr) [reply avm/res/document-db/database-account/gremlin-database/graph/main.json:1]: Interesting, ... that did nothing. I could've sworn this worked in the past.

File: avm/utilities/pipelines/staticValidation/psrule/ps-rule.yaml [hunk 2]
```diff
@@ -81,3 +81,4 @@ rule:
     - Azure.VM.UseHybridUseBenefit
     - Azure.Storage.UseReplication
     - Azure.AppConfig.PurgeProtect
+    - Custom.Azure.Storage.UseReplication # Custom rule not working as expected
```
[the remark this answers is not in the document]
Comment (Agazoth) [reply avm/utilities/pipelines/staticValidation/psrule/ps-rule.yaml]: I have removed Custom.Azure.Storage.UseReplication from the exceptoion rules and updated the custom rule to include the correct sku name:
image

That seems to do the trick.

File: avm/res/storage/storage-account/README.md [hunk 3]
```diff
@@ -19,6 +19,7 @@ This module deploys a Storage Account.
 | `Microsoft.Authorization/locks` | [2020-05-01](https://learn.microsoft.com/en-us/azure/templates/Microsoft.Authorization/2020-05-01/locks) |
 | `Microsoft.Authorization/roleAssignments` | [2022-04-01](https://learn.microsoft.com/en-us/azure/templates/Microsoft.Authorization/2022-04-01/roleAssignments) |
 | `Microsoft.Insights/diagnosticSettings` | [2021-05-01-preview](https://learn.microsoft.com/en-us/azure/templates/Microsoft.Insights/2021-05-01-preview/diagnosticSettings) |
+| `Microsoft.KeyVault/vaults/secrets` | [2022-07-01](https://learn.microsoft.com/en-us/azure/templates/Microsoft.KeyVault/2022-07-01/vaults/secrets) |
```
Comment (eriqua) [inline avm/res/storage/storage-account/README.md:22]: Same comment of the cognitive PR https://github.com/Azure/bicep-registry-modules/pull/1932#pullrequestreview-2056624383

Comment (Agazoth) [reply avm/res/storage/storage-account/README.md:22]: I absolutely agree with your comment on #1932 - i have startet this discussion: #1934

**Comment by jikuja** [comment]
How does this work on following scenarios:

* re-deployment
* re-deployment after storage account key rotation?

**Comment by AlexanderSehr** [comment]
> How does this work on following scenarios:
> 
> * re-deployment
> * re-deployment after storage account key rotation?

Presumably, all secrets would be overwritten/updated to latest each time the deployment runs

**Comment by AlexanderSehr** [comment]
Hey @Agazoth,
for reference (I may not have mentioned it before), we documented the interface [here](https://azure.github.io/Azure-Verified-Modules/specs/bicep/interfaces/#secrets-export). 💪

**Comment by Agazoth** [comment]
> Hey @Agazoth, for reference (I may not have mentioned it before), we documented the interface [here](https://azure.github.io/Azure-Verified-Modules/specs/bicep/interfaces/#secrets-export). 💪 Also updated DocumentDB (the original first module with the feature) since then.

Thanks a bunck @AlexanderSehr - I have updated the code by following your documentation - it woerked like a charm :-)

It should be ready for production now.

File: avm/res/storage/storage-account/main.bicep [hunk 4]
```diff
@@ -620,6 +623,52 @@ module storageAccount_tableServices 'table-service/main.bicep' = if (!empty(tabl
...
+  name: '${uniqueString(deployment().name, location)}-secrets-kv'
+  scope: resourceGroup(
+    split((secretsExportConfiguration.?keyVaultResourceId ?? '//'), '/')[2],
+    split((secretsExportConfiguration.?keyVaultResourceId ?? '////'), '/')[4]
+  )
+  params: {
+    keyVaultName: last(split(secretsExportConfiguration.?keyVaultResourceId ?? '//', '/'))
+    secretsToSet: union(
+      [],
+      contains(secretsExportConfiguration!, 'Key1Key')
```
Comment (AlexanderSehr) [inline avm/res/storage/storage-account/main.bicep]: I'd probably suggest to call it something like `accessKey1`. What was the idea behind `Key*Key`?

Comment (AlexanderSehr) [reply avm/res/storage/storage-account/main.bicep]: Similar ask for the connectionString

File: avm/res/storage/storage-account/main.bicep [hunk 5]
```diff
@@ -628,6 +631,52 @@ module storageAccount_tableServices 'table-service/main.bicep' = if (!empty(tabl
...
+  name: '${uniqueString(deployment().name, location)}-secrets-kv'
+  scope: resourceGroup(
+    split((secretsExportConfiguration.?keyVaultResourceId ?? '//'), '/')[2],
+    split((secretsExportConfiguration.?keyVaultResourceId ?? '////'), '/')[4]
+  )
+  params: {
+    keyVaultName: last(split(secretsExportConfiguration.?keyVaultResourceId ?? '//', '/'))
+    secretsToSet: union(
+      [],
+      contains(secretsExportConfiguration!, 'key1')
```
Comment (AlexanderSehr) [inline avm/res/storage/storage-account/main.bicep:644]:
```suggestion
      contains(secretsExportConfiguration!, 'accessKey1')
```

File: avm/res/storage/storage-account/main.bicep [hunk 6]
```diff
@@ -628,6 +631,52 @@ module storageAccount_tableServices 'table-service/main.bicep' = if (!empty(tabl
...
+        : [],
+      contains(secretsExportConfiguration!, 'connectionString1')
+        ? [
+            {
+              name: secretsExportConfiguration!.connectionString1
+              value: 'DefaultEndpointsProtocol=https;AccountName=${storageAccount.name};AccountKey=${storageAccount.listKeys().keys[0]};EndpointSuffix=core.windows.net'
+            }
+          ]
+        : [],
+      contains(secretsExportConfiguration!, 'key2')
```
Comment (AlexanderSehr) [inline avm/res/storage/storage-account/main.bicep:660]:
```suggestion
      contains(secretsExportConfiguration!, 'accessKey2')
```

File: avm/res/storage/storage-account/main.bicep (hunk 5 above)
Comment (AlexanderSehr) [reply avm/res/storage/storage-account/main.bicep:644]: Would need update in other places too

File: avm/res/storage/storage-account/main.bicep (hunk 6 above)
Comment (AlexanderSehr) [reply avm/res/storage/storage-account/main.bicep:660]: Would need update in other places too

File: avm/res/storage/storage-account/tests/e2e/kvSecrets/dependencies.bicep [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+      family: 'A'
+      name: 'standard'
+    }
+    enableRbacAuthorization: true
+    tenantId: subscription().tenantId
+  }
+}
+
+@description('The name of the Key Vault created.')
+output keyVaultId string = keyVault.id
```
Comment (AlexanderSehr) [inline avm/res/storage/storage-account/tests/e2e/kvSecrets/dependencies.bicep:21]:
```suggestion
output keyVaultResouceId string = keyVault.id
```

File: avm/res/storage/storage-account/main.bicep [hunk 8]
```diff
@@ -906,13 +906,13 @@ type secretsExportConfigurationType = {
   keyVaultResourceId: string
 
   @description('Optional. Default to Key1Key. The primary key secret name to create.')
```
Comment (AlexanderSehr) [inline avm/res/storage/storage-account/main.bicep]: Please also update each description

**Review by Agazoth — COMMENTED** [review]
Already done

**Comment by AlexanderSehr** [comment]
Nice, looks good to me. @fblix / @ktremain any doubts from your side?

**Comment by ktremain** [comment]
@Agazoth can you please update this again from the current main and bump the version, as a recent commit now has a matching version number as your code.

**Comment by Agazoth** [comment]
@ktremain done :-)

**Comment by ktremain** [comment]
@Agazoth Thanks for updating the branch, please can you also increase the version number in version.json (same folder as the main.bicep), otherwise the back-end automation wont see it as a new release to publish to the Bicep module registry.

**Comment by Agazoth** [comment]
@ktremain sorry about that 😄 - now I'm done ...

**Comment by ktremain** [comment]
@Azure/alz-core-team - Erika added the Do not merge label previously - is it ok to remove this now? the PR looks good to me.

**Comment by AlexanderSehr** [comment]
It's good now. The label was previously added because the schema was not defined & [documented](https://azure.github.io/Azure-Verified-Modules/specs/bicep/interfaces/#secrets-export) yet.

On that note @ktremain, just two quick comments on what you mentioned
- 'Azure/alz-core-team' is not related to AVM
- Any PR that contains changes to module files (most notably the `main.json`) will result in a new version. However, as it would just be a new patch/bug version (<major>.<minor>.<patch>) I agree that it makes sense to update the <minor> in this case

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**